File: ex/clean/src/main/kotlin/com/ktb/shop/adapters/web/CustomerController.kt
package com.ktb.shop.adapters.web

import com.ktb.shop.usecases.register.RegisterCustomer
import com.ktb.shop.usecases.register.RegisterInput
import io.micronaut.http.HttpResponse
import io.micronaut.http.MediaType
import io.micronaut.http.annotation.Body
import io.micronaut.http.annotation.Controller
import io.micronaut.http.annotation.Post
import io.swagger.v3.oas.annotations.Operation
import io.swagger.v3.oas.annotations.media.Content
import io.swagger.v3.oas.annotations.media.Schema
import io.swagger.v3.oas.annotations.responses.ApiResponse
import io.swagger.v3.oas.annotations.responses.ApiResponses
import io.swagger.v3.oas.annotations.tags.Tag
import jakarta.validation.Valid

// INTERFACE ADAPTER (driving). Depends only on the INPUT boundary (RegisterCustomer),
// never on the interactor. Maps web DTO <-> use-case models. No business rules here.
// `open` because Micronaut applies validation AOP to `register`, and KSP requires
// the advised method to be non-final (the kotlin all-open plugin runs too late for KSP).
@Controller("/api/customers")
@Tag(name = "Customers")
open class CustomerController(private val registerCustomer: RegisterCustomer) {

    @Post
    @Operation(
        summary = "สมัครสมาชิกใหม่",
        description = "สร้างลูกค้าใหม่ — entity (Email, CustomerId) ตรวจความถูกต้องที่แก่นกลาง; " +
            "ถ้าอีเมลซ้ำจะได้ 409, ถ้า payload ผิดรูปจะได้ 400 โดยไม่แตะ DB",
    )
    @ApiResponses(
        ApiResponse(
            responseCode = "201",
            description = "สมัครสำเร็จ",
            content = [Content(mediaType = MediaType.APPLICATION_JSON, schema = Schema(implementation = CustomerResponse::class))],
        ),
        ApiResponse(
            responseCode = "400",
            description = "payload ผิดรูป (validation ล้มเหลว / อีเมลผิดรูปแบบ)",
            content = [Content(mediaType = MediaType.APPLICATION_JSON, schema = Schema(implementation = ApiError::class))],
        ),
        ApiResponse(
            responseCode = "409",
            description = "อีเมลถูกใช้ลงทะเบียนไปแล้ว",
            content = [Content(mediaType = MediaType.APPLICATION_JSON, schema = Schema(implementation = ApiError::class))],
        ),
        ApiResponse(
            responseCode = "429",
            description = "เรียกถี่เกิน rate limit (100 req/นาที ต่อ IP)",
            content = [Content(mediaType = MediaType.APPLICATION_JSON, schema = Schema(implementation = ApiError::class))],
        ),
    )
    open fun register(@Valid @Body req: RegisterRequest): HttpResponse<CustomerResponse> {
        val out = registerCustomer.register(RegisterInput(req.id, req.email, req.name))
        return HttpResponse.created(CustomerResponse(out.id, out.email, out.name))
    }
}

File: ex/clean-go/internal/adapters/web/customer_controller.go
package web

import (
	"encoding/json"
	"net/http"

	"github.com/ktb/clean-shop/internal/usecases/register"
)

// INTERFACE ADAPTER (driving). Depends only on the INPUT boundary
// (register.RegisterCustomer), never on the concrete interactor. Maps web
// DTO <-> use-case models. No business rules here.
type CustomerController struct {
	registerCustomer register.RegisterCustomer
}

func NewCustomerController(registerCustomer register.RegisterCustomer) *CustomerController {
	return &CustomerController{registerCustomer: registerCustomer}
}

// POST /api/customers
func (c *CustomerController) Register(w http.ResponseWriter, r *http.Request) {
	var req RegisterRequest
	if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
		writeJSON(w, http.StatusBadRequest, APIError{Code: "bad_request", Message: "invalid input"})
		return
	}
	if err := req.Validate(); err != nil {
		writeJSON(w, http.StatusBadRequest, APIError{Code: "bad_request", Message: "invalid input"})
		return
	}

	out, err := c.registerCustomer.Register(r.Context(), register.RegisterInput{
		ID:    req.ID,
		Email: req.Email,
		Name:  req.Name,
	})
	if err != nil {
		writeError(w, err)
		return
	}
	writeJSON(w, http.StatusCreated, CustomerResponse{
		ID:    out.ID,
		Email: out.Email,
		Name:  out.Name,
	})
}
